flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerService.java

```diff
@@ -18,43 +18,306 @@
+	 * @param stream the stream to read from.
+	 * @param keyGroupIdx the id of the key-group to be put in the snapshot.
+	 * @param userCodeClassLoader the class loader that will be used to deserialize
+	 * 								the local key and namespace serializers.
+	 */
+	public void restoreTimersForKeyGroup(DataInputViewStreamWrapper stream, int keyGroupIdx, ClassLoader userCodeClassLoader) throws IOException, ClassNotFoundException {
+		TypeSerializer<K> tmpKeySerializer = InstantiationUtil.deserializeObject(stream, userCodeClassLoader);
+		TypeSerializer<N> tmpNamespaceSerializer = InstantiationUtil.deserializeObject(stream, userCodeClassLoader);
+
+		if ((this.keySerializer != null && !this.keySerializer.equals(tmpKeySerializer)) ||
+					(this.namespaceSerializer != null && !this.namespaceSerializer.equals(tmpNamespaceSerializer))) {
+
+				throw new IllegalArgumentException("Tried to restore timers " +
+						"for the same service with different serializers.");
+		}
+
+		this.keySerializer = tmpKeySerializer;
+		this.namespaceSerializer = tmpNamespaceSerializer;
+
+		InternalTimer.TimerSerializer<K, N> timerSerializer =
+				new InternalTimer.TimerSerializer<>(this.keySerializer, this.namespaceSerializer);
+
+		checkArgument(keyGroupRange.contains(keyGroupIdx),
+				"Key Group " + keyGroupIdx + " does not belong to the local range.");
+
+		// read the event time timers
+		int sizeOfEventTimeTimers = stream.readInt();
+		if (sizeOfEventTimeTimers > 0) {
+			List<InternalTimer<K, N>> eventTimeTimers = new ArrayList<>();
+			for (int i = 0; i < sizeOfEventTimeTimers; i++) {
+				InternalTimer<K, N> timer = timerSerializer.deserialize(stream);
+				
+				eventTimeTimers.add(timer);
+			}
+
+			restoreEventTimeTimersForKeyGroup(keyGroupIdx, eventTimeTimers);
+		}
+
+		// read the processing time timers
+		int sizeOfProcessingTimeTimers = stream.readInt();
+		if (sizeOfProcessingTimeTimers > 0) {
+			List<InternalTimer<K, N>> processingTimeTimers = new ArrayList<>();
+			for (int i = 0; i < sizeOfProcessingTimeTimers; i++) {
+				InternalTimer<K, N> timer = timerSerializer.deserialize(stream);
+				processingTimeTimers.add(timer);
+			}
+
+			restoreProcessingTimeTimersForKeyGroup(keyGroupIdx, processingTimeTimers);
+		}
+	}
+
+	/**
+	 * Starts the local {@link InternalTimerService} by:
+	 * <ol>
+	 *     <li>Setting the {@code keySerialized} and {@code namespaceSerializer} for the timers it will contain.</li>
+	 *     <li>Setting the {@code triggerTarget} which contains the action to be performed when a timer fires.</li>
+	 *     <li>Re-registering timers that were retrieved after recoveting from a node failure, if any.</li>
+	 * </ol>
+	 * This method can be called multiple times, as long as it is called with the same serializers.
+	 */
+	void startTimerService(
+			TypeSerializer<K> keySerializer,
+			TypeSerializer<N> namespaceSerializer,
+			Triggerable<K, N> triggerTarget)
+	{
+
+		if (isInitialized) {
```

Comment:
Either `isInitialized` does not require to be volatile or this code is potentially broken. If you need the thread-safety, I suggest replacing this with `AtomicBoolean::compareAndSet(...)`.